nano/core_test/peer_container.cpp

```diff
@@ -11,71 +19,100 @@ TEST (peer_container, empty_peers)
+	node_flags.disable_rep_crawler = true;
+	auto & node1 = *system.add_node (node_config, node_flags);
+	auto outer_node1 = nano::test::add_outer_node (system, nano::test::get_available_port (), node_flags);
+	outer_node1->config.network_params.network.keepalive_period = std::chrono::minutes (10);
+	auto outer_node2 = nano::test::add_outer_node (system, nano::test::get_available_port (), node_flags);
+	outer_node2->config.network_params.network.keepalive_period = std::chrono::minutes (10);
+	auto now = std::chrono::steady_clock::now ();
+	auto channel1 = nano::test::establish_tcp (system, node1, outer_node1->network.endpoint ());
 	ASSERT_NE (nullptr, channel1);
-	node1.network.udp_channels.modify (channel1, [&now] (auto channel) {
-		channel->set_last_packet_received (now - std::chrono::seconds (1));
+	// set the last packet sent for channel1 only to guarantee it contains a value.
+	// it won't be necessarily the same use by the cleanup cutoff time
+	node1.network.tcp_channels.modify (channel1, [&now] (auto channel) {
+		channel->set_last_packet_sent (now - std::chrono::seconds (5));
 	});
-	auto channel2 (node1.network.udp_channels.insert (endpoint2, 0));
+	auto channel2 = nano::test::establish_tcp (system, node1, outer_node2->network.endpoint ());
 	ASSERT_NE (nullptr, channel2);
-	node1.network.udp_channels.modify (channel2, [&now] (auto channel) {
-		channel->set_last_packet_received (now + std::chrono::seconds (1));
+	// set the last packet sent for channel2 only to guarantee it contains a value.
+	// it won't be necessarily the same use by the cleanup cutoff time
+	node1.network.tcp_channels.modify (channel2, [&now] (auto channel) {
+		channel->set_last_packet_sent (now + std::chrono::seconds (1));
 	});
 	ASSERT_EQ (2, node1.network.size ());
-	ASSERT_EQ (2, node1.network.udp_channels.size ());
-	node1.network.cleanup (now);
+	ASSERT_EQ (2, node1.network.tcp_channels.size ());
+
+	for (auto it = 0; node1.network.tcp_channels.size () > 1 && it < 10; ++it)
+	{
+		// we can't control everything the nodes are doing in background, so using the middle time as
+		// the cutoff point.
+		auto const channel1_last_packet_sent = channel1->get_last_packet_sent ();
+		auto const channel2_last_packet_sent = channel2->get_last_packet_sent ();
+		auto const max_last_packet_sent = std::max (channel1_last_packet_sent, channel2_last_packet_sent);
+		auto const min_last_packet_sent = std::min (channel1_last_packet_sent, channel2_last_packet_sent);
+		auto const cleanup_point = ((max_last_packet_sent - min_last_packet_sent) / 2) + min_last_packet_sent;
+
+		node1.network.cleanup (cleanup_point);
+
+		// it is possible that the last_packet_sent times changed because of another thread and the cleanup_point
+		// is not the middle time anymore, in these case we wait a bit and try again in a loop up to 10 times
+		if (node1.network.tcp_channels.size () == 2)
+		{
+			std::this_thread::sleep_for (std::chrono::milliseconds (500));
```

Comment:
Use the WAIT macro to sleep. It is important to keep the calling the system poll function for background processes to continue working and WAIT() dos that.